### PR TITLE
Keep `LogCoshError` finite for large residuals

### LINKED ISSUE #3488: LogCoshError returns inf for finite large residuals
`log_cosh_error` currently evaluates the exponential expression directly, which can overflow for finite residuals.

```python
import torch
from torchmetrics.functional.regression import log_cosh_error

preds = torch.tensor([100.0, -100.0])
target = torch.zeros_like(preds)

print(log_cosh_error(preds, target))  # tensor(inf)
```

The expected result is approximately `99.30685`. Float16 inputs can also overflow during subtraction or accumulation.

This reproduces on current `master` (`8d008de1`) with Python 3.12.13 and PyTorch 2.13.0. I have a small fix and regression tests ready :)

### PR DESCRIPTION
**Pull request by Robertboy18**
## What does this PR do?

Fixes #3488

I was checking `LogCoshError` with larger residuals and noticed the direct exponential expression can return `inf` even when the metric is finite. Half-precision inputs can also overflow during subtraction or accumulation.

This uses the equivalent `logaddexp(diff, -diff) - log(2)` expression and promotes float16/bfloat16 inputs before subtraction. I added regression coverage for both APIs across float16, bfloat16, float32, and float64.

The full LogCosh test file passes, along with pre-commit, TorchScript, `torch.compile`, gradient, and CUDA checks. I can also add a small Lean/TorchLean proof of the equivalence if useful :)

### CODE REVIEW

**Review by justusschock — APPROVED** [review]

File: src/torchmetrics/functional/regression/log_cosh.py [hunk 1]
```diff
@@ -43,8 +45,14 @@ def _log_cosh_error_update(preds: Tensor, target: Tensor, num_outputs: int) -> t
...
     preds, target = _unsqueeze_tensors(preds, target)
+    if preds.dtype == torch.float16 or preds.dtype == torch.bfloat16:
+        preds = preds.float()
+    if target.dtype == torch.float16 or target.dtype == torch.bfloat16:
+        target = target.float()
     diff = preds - target
-    sum_log_cosh_error = torch.log((torch.exp(diff) + torch.exp(-diff)) / 2).sum(0).squeeze()
+    if not diff.is_floating_point():
+        diff = diff.float()
+    sum_log_cosh_error = (torch.logaddexp(diff, -diff) - math.log(2.0)).sum(0).squeeze()
```
[the remark this answers is not in the document]
Comment (Robertboy18) [reply src/torchmetrics/functional/regression/log_cosh.py:57]: Thanks, the complex case is valid. Integer inputs did work previously because `torch.exp` promotes them, so I kept that behavior. Complex inputs are now rejected explicitly in both APIs, with regression tests for both cases :)

**Review by Borda — APPROVED** [review]
